lib/gcs/index.js

```diff
@@ -0,0 +1,93 @@
+'use strict';
+
+const fs = require('fs-extra');
+const path = require('path');
+const readdir = require('recursive-readdir');
+// Documentation of @google-cloud/storage: https://cloud.google.com/nodejs/docs/reference/storage/2.3.x/Bucket#upload
+const { Storage } = require('@google-cloud/storage');
+
+const log = require('intel').getLogger('sitespeedio.plugin.gcs');
+const throwIfMissing = require('../../support/util').throwIfMissing;
+
+async function upload(dir, gcsOptions, prefix) {
+  const files = await readdir(dir);
+  const promises = [];
+
+  const storage = new Storage({
+    projectId: gcsOptions.projectId,
+    keyFilename: gcsOptions.key
+  });
+
+  const bucket = storage.bucket(gcsOptions.bucketname);
+
+  for (let file of files) {
+    promises.push(uploadFile(file, bucket, gcsOptions, prefix, dir));
+  }
+  return Promise.all(promises);
+}
+
+async function uploadFile(file, bucket, gcsOptions, prefix, baseDir) {
+  const subPath = path.relative(baseDir, file);
+  const fileName = path.join(gcsOptions.path || prefix, subPath);
+
+  const params = {
+    public: !!gcsOptions.public,
+    destination: fileName,
+    resumable: true,
+    validation: 'crc32c',
+    gzip: true,
+    metadata: {
+      metadata: {
+        cacheControl: 'public, max-age=31536000'
+      }
+    }
+  };
+
+  await bucket.upload(file, params);
```

Comment:
Small thing: here you can just return the promise and the user of the function will choose to wait/then etc.

```suggestion
  return bucket.upload(file, params);
```